gobblin-runtime/src/main/java/gobblin/runtime/TaskStateCollectorService.java

```diff
@@ -115,30 +115,32 @@ protected void shutDown() throws Exception {
    * @throws IOException if it fails to collect the output {@link TaskState}s
    */
   private void collectOutputTaskStates() throws IOException {
-    if (!this.fs.exists(this.outputTaskStateDir)) {
-      LOGGER.warn(String.format("Output task state path %s does not exist", this.outputTaskStateDir));
-      return;
-    }
-
-    FileStatus[] fileStatuses = this.fs.listStatus(this.outputTaskStateDir, new PathFilter() {
+    List<String> taskStateNames = taskStateStore.getStateNames(outputTaskStateDir.getName(), new Predicate<String>() {
       @Override
-      public boolean accept(Path path) {
-        return path.getName().endsWith(AbstractJobLauncher.TASK_STATE_STORE_TABLE_SUFFIX)
-            && !path.getName().startsWith(FsStateStore.TMP_FILE_PREFIX);
-      }
-    });
-    if (fileStatuses == null || fileStatuses.length == 0) {
+      public boolean apply(String input) {
+        return input.endsWith(AbstractJobLauncher.TASK_STATE_STORE_TABLE_SUFFIX)
+        && !input.startsWith(FsStateStore.TMP_FILE_PREFIX);
+      }});
+
+    if (taskStateNames == null || taskStateNames.size() == 0) {
       LOGGER.warn("No output task state files found in " + this.outputTaskStateDir);
       return;
     }
 
-    Queue<TaskState> taskStateQueue = Queues.newConcurrentLinkedQueue();
-    try (ParallelRunner stateSerDeRunner = new ParallelRunner(this.stateSerDeRunnerThreads, this.fs)) {
-      for (FileStatus status : fileStatuses) {
-        LOGGER.debug("Found output task state file " + status.getPath());
+    final Queue<TaskState> taskStateQueue = Queues.newConcurrentLinkedQueue();
+    try (ParallelRunner stateSerDeRunner = new ParallelRunner(this.stateSerDeRunnerThreads, null)) {
+      for (final String taskStateName : taskStateNames) {
+        LOGGER.debug("Found output task state file " + taskStateName);
         // Deserialize the TaskState and delete the file
-        stateSerDeRunner.deserializeFromSequenceFile(Text.class, TaskState.class, status.getPath(), taskStateQueue,
-            true);
+        stateSerDeRunner.submitCallable(new Callable<Void>() {
+          @Override
+          public Void call() throws Exception {
+            TaskState taskState = taskStateStore.getAll(outputTaskStateDir.getName(), taskStateName).get(0);
+            taskStateQueue.add(taskState);
+            taskStateStore.delete(outputTaskStateDir.getName(), taskStateName);
```

Comment:
This is a bit risky because certain key-value structures (for example an in-memory map) do not allow concurrent modification. Ideally, the delete should be done after the loop has completed.